pkg/updatestrategy/aws_asg.go

```diff
@@ -339,23 +340,32 @@ func (n *ASGNodePoolsBackend) Terminate(node *Node, decrementDesired bool) error
 		ShouldDecrementDesiredCapacity: aws.Bool(decrementDesired),
 	}
 
-	_, err := n.asgClient.TerminateInstanceInAutoScalingGroup(params)
-	if err != nil {
-		log.Errorf("Faild to terminate instance '%s': %v", instanceId, err)
-		_, serr := n.instanceState(instanceId)
-		if serr != nil {
-			return fmt.Errorf("failed to terminate instance '%s': %v, %v", instanceId, err, serr)
-		}
-	}
-
-	// wait for the instance to be terminated/stopped
-	instanceState := func() error {
+	terminateAsgInstance := func() error {
 		state, err := n.instanceState(instanceId)
 		if err != nil {
 			return backoff.Permanent(err)
 		}
-
 		switch state {
+		// if instance is running terminate it
+		case ec2.InstanceStateNameRunning:
+			_, err = n.asgClient.TerminateInstanceInAutoScalingGroup(params)
+			if err != nil {
+				if aerr, ok := err.(awserr.Error); ok {
+					switch aerr.Code() {
+					// if an operation is in progress then retry later.
+					case autoscaling.ErrCodeScalingActivityInProgressFault, autoscaling.ErrCodeResourceContentionFault:
+						return errors.New("waiting for AWS to complete operations")
+					// otherwise fail
+					default:
+						return backoff.Permanent(err)
+					}
+				} else {
+					// at this point the call to the API failed. try later
+					return err
+				}
+			} else {
+				return errors.New("signalled instance to shutdown")
+			}
```

Comment:
No need for the two `else` since you return in the if in both cases :)

Reply:
The reason I wrote it like that is so that there are different errors returns for different situations. I thought it would be good debugging information. But turns out the `backoff` library doesn't ever log anything so it doesn't matter. Will change.